### PR TITLE
[Issue #249] - Add support of loading CasC definition from WAR file bundles

### PR DESCRIPTION
**Pull request by oleg-nenashev**
Fix for #249. I am working on a [Custom WAR Packager](https://github.com/jenkinsci/custom-war-packager) project, which allows bundling Jenkins to a portable WAR with all configurations and plugins. Currently it supports System Groovy Scripts, but it would be great to also support Configuration-as-Code as input. Tracking ticket: [JENKINS-51656](https://issues.jenkins-ci.org/browse/JENKINS-51656)

In order to do that, CasC plugin should check WAR resources while loading configs. I propose to use layout similar to what Groovy hooks do:

* `WEB-INF/casc.yaml`
* `WEB-INF/casc.yaml.d/*`

Reference implementation: https://github.com/jenkinsci/custom-war-packager/pull/30

@ndeloof @ewelinawilkosz @MadsNielsen

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java [hunk 1]
```diff
@@ -136,24 +143,67 @@ public static void init() throws Exception {
...
+        try {
+            URL bundled = servletContext.getResource("/WEB-INF/"+ cascFile);
+            if (bundled != null) {
+                res.add(bundled.toString());
+            }
+        } catch (IOException e) {
+            LOGGER.log(WARNING, "Failed to load /WEB-INF/" + cascFile, e);
+        }
+
+        Set<String> resources = servletContext.getResourcePaths("/WEB-INF/"+ cascDirectory +"/");
```
Comment (MadsNielsen) [inline src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java]: Would it make sense to only look for `.yml` or `.yaml` files here? Are there any situations where people might end up putting non yaml files in this location?

File: src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java [hunk 2]
```diff
@@ -136,24 +143,67 @@ public static void init() throws Exception {
...
+        }
+        if (configParameters.isEmpty()) {
+            LOGGER.log(Level.FINE, "No configuration set nor default config file");
+            return;
+        }
+        configure(configParameters);
+    }
+
+    public List<String> getBundledCasCURIs() {
+        final String cascFile = "casc.yml";
```
Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java]: we used `jenkins.yaml` as default configuration file name, should be the same

File: src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java (hunk 1 above)
Comment (ndeloof) [reply src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java]: Agree, should filter by YAML_FILES_PATTERN as we do for a fielsystem-based conf.d

File: src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java [hunk 3]
```diff
@@ -136,24 +143,67 @@ public static void init() throws Exception {
...
+        final String cascDirectory = "casc.yml.d";
```
Comment (ndeloof) [inline src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java]: I don't have great suggestion but don't really like `casc.yml.d` as standard directory name. `configuration.d`?

File: src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java (hunk 1 above)
Comment (oleg-nenashev) [reply src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java]: Agreed

File: src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java (hunk 2 above)
Comment (oleg-nenashev) [reply src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java]: works for me

File: src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java (hunk 3 above)
Comment (oleg-nenashev) [reply src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java]: Or jenkins.yaml.d to keep it aligned?

Comment (ndeloof) [reply src/main/java/org/jenkinsci/plugins/casc/ConfigurationAsCode.java]: good point, LGTM

**Comment by oleg-nenashev** [comment]
@ndeloof @MadsNielsen comments addressed

**Comment by ndeloof** [comment]
LGTM. Can you please rebase on master's HEAD so we can merge ?

**Comment by oleg-nenashev** [comment]
@ndeloof You can just squash and merge it from UI. In such case branch things will dissapear

**Pull request merged**